# Pandas - Ejercicios dataset Titanic



- - -

In [81]:
import pandas as pd

In [82]:
df = pd.read_csv("https://raw.githubusercontent.com/4data-lab/datasets/master/titanic.csv")

#### Columnas
survived (Sobrevivió)  
pclass (Clase de números)  
sex (Sexo)  
age (Edad)  
sibsp (Número de hermanos/cónyuges abordo)  
parch (Número de padres/hijos)   
fare (Tarifa)  
embarked (Embarcado)  
class (Clase en texto)  
adult_male (Hombre Adulto)  
deck (Cubierta)  
embark_town (Pueblo de embarcación)  
alive (Vivo)  
alone (Solo)


- - -
1 - Muestra solo los primeros 5 registros

In [83]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

- - -
2 - Obtén cuales son las distintas clases de viaje de este dataframe

`array(['Third', 'First', 'Second'], dtype=object)`

In [84]:
df.dtypes # 'Third', 'First', 'Second' son valores de la columa "class" que vemos que es de tipo "object"

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object

In [85]:
df["class"].unique() # Devuelve los valores únicos de la columna "class" y su tipo. ## ES LO PEDIDO EN EJERCICIO 2

array(['Third', 'First', 'Second'], dtype=object)

- - -
3 - Obten el valor máximo de edad en el dataframe

`Out: 80.0`

In [86]:
df["age"].max()

80.0

- - -
4 - Obten cúantos valores distintos hay en la columna de Pueblo de embarcación

`Out: 3`

In [87]:
df["embarked"].count() # Cuantos valores embarked hay (o sea valores que no sean missing values)

889

In [88]:
df["embarked"].unique() # Cuales son los valores de embarked úncios? Vemos que son 3: S, C, Q

array(['S', 'C', 'Q', nan], dtype=object)

In [89]:
df["embarked"].nunique() # Cuantos valores de embarked diferentes hay?  ## ES LO PEDIDO EN EJERCICIO 4

3

- - -
5 - Muestra solo las primeras cinco edades más frencuentes

`Out:
24.0    30
22.0    27
18.0    26
19.0    25
30.0    25
Name: age, dtype: int64`

In [90]:
df["age"].value_counts() # frecuencia absoluta de todos los valores

24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: age, Length: 88, dtype: int64

In [91]:
df["age"].value_counts().head() # frecuencia absoluta de solo los 5 primeros valores

24.0    30
22.0    27
18.0    26
19.0    25
28.0    25
Name: age, dtype: int64

In [92]:
list(df["age"].value_counts().head()) # Para tener la lista

[30, 27, 26, 25, 25]

In [93]:
 # Para tener el Out pedido en el Ejercicio
string=""
for i in list(df["age"].value_counts().head()):
  string += str(i)+ ".0 "
print("Out: " + string + " Name: age, dtype: int64")

Out: 30.0 27.0 26.0 25.0 25.0  Name: age, dtype: int64


- - -
6 - Muestra los registros que cumplan la condición de disponer el valor de cubierta ("deck") "F"

In [94]:
df[df["deck"] == "F"]

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
66          1       2  female  29.0      0      0  10.5000        S  Second   
75          0       3    male  25.0      0      0   7.6500        S   Third   
128         1       3  female   NaN      1      1  22.3583        C   Third   
148         0       2    male  36.5      0      2  26.0000        S  Second   
183         1       2    male   1.0      2      1  39.0000        S  Second   
193         1       2    male   3.0      1      1  26.0000        S  Second   
340         1       2    male   2.0      1      1  26.0000        S  Second   
345         1       2  female  24.0      0      0  13.0000        S  Second   
516         1       2  female  34.0      0      0  10.5000        S  Second   
618         1       2  female   4.0      2      1  39.0000        S  Second   
699         0       3    male  42.0      0      0   7.6500        S   Third   
715         0       3    male  19.0      0      0   7.6500        S   Third   
776         0       3    male   NaN      0      0   7.7500        Q   Third   

     adult_male deck  embark_town alive  alone  
66        False    F  Southampton   yes   True  
75         True    F  Southampton    no   True  
128       False    F    Cherbourg   yes  False  
148        True    F  Southampton    no  False  
183       False    F  Southampton   yes  False  
193       False    F  Southampton   yes  False  
340       False    F  Southampton   yes  False  
345       False    F  Southampton   yes   True  
516       False    F  Southampton   yes   True  
618       False    F  Southampton   yes  False  
699        True    F  Southampton    no   True  
715        True    F  Southampton    no   True  
776        True    F   Queenstown    no   True

In [95]:
df["deck"] == "F"

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: deck, Length: 891, dtype: bool

- - -
7 - Muestra los registros que cumplan la condición de disponer más de 47 años y menos de 49 años

In [96]:
df[(df["age"] > 47) & (df["age"] < 49)] #suponemos que el ejercicio pide más estricatmente y menos estrictamente

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
460         1       1    male  48.0      0      0  26.5500        S   First   
463         0       2    male  48.0      0      0  13.0000        S  Second   
556         1       1  female  48.0      1      0  39.6000        C   First   
645         1       1    male  48.0      1      0  76.7292        C   First   
712         1       1    male  48.0      1      0  52.0000        S   First   
736         0       3  female  48.0      1      3  34.3750        S   Third   
754         1       2  female  48.0      1      2  65.0000        S  Second   
771         0       3    male  48.0      0      0   7.8542        S   Third   
862         1       1  female  48.0      0      0  25.9292        S   First   

     adult_male deck  embark_town alive  alone  
460        True    E  Southampton   yes   True  
463        True  NaN  Southampton    no   True  
556       False    A    Cherbourg   yes  False  
645        True    D    Cherbourg   yes  False  
712        True    C  Southampton   yes  False  
736       False  NaN  Southampton    no  False  
754       False  NaN  Southampton   yes  False  
771        True  NaN  Southampton    no   True  
862       False    D  Southampton   yes   True

In [97]:
 df[df["age"] == 48] # Relamente lo anterior es mismo que decir = 48

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
460         1       1    male  48.0      0      0  26.5500        S   First   
463         0       2    male  48.0      0      0  13.0000        S  Second   
556         1       1  female  48.0      1      0  39.6000        C   First   
645         1       1    male  48.0      1      0  76.7292        C   First   
712         1       1    male  48.0      1      0  52.0000        S   First   
736         0       3  female  48.0      1      3  34.3750        S   Third   
754         1       2  female  48.0      1      2  65.0000        S  Second   
771         0       3    male  48.0      0      0   7.8542        S   Third   
862         1       1  female  48.0      0      0  25.9292        S   First   

     adult_male deck  embark_town alive  alone  
460        True    E  Southampton   yes   True  
463        True  NaN  Southampton    no   True  
556       False    A    Cherbourg   yes  False  
645        True    D    Cherbourg   yes  False  
712        True    C  Southampton   yes  False  
736       False  NaN  Southampton    no  False  
754       False  NaN  Southampton   yes  False  
771        True  NaN  Southampton    no   True  
862       False    D  Southampton   yes   True

In [98]:
df[(df["age"] >= 47) & (df["age"] <= 49)] # Si no queremos hacer la condición de estrictamente menor y estrictamente mayor, sino que también aceptamos los iguales sería así

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
52          1       1  female  49.0      1      0   76.7292        C   First   
110         0       1    male  47.0      0      0   52.0000        S   First   
132         0       3  female  47.0      1      0   14.5000        S   Third   
453         1       1    male  49.0      1      0   89.1042        C   First   
460         1       1    male  48.0      0      0   26.5500        S   First   
462         0       1    male  47.0      0      0   38.5000        S   First   
463         0       2    male  48.0      0      0   13.0000        S  Second   
515         0       1    male  47.0      0      0   34.0208        S   First   
556         1       1  female  48.0      1      0   39.6000        C   First   
586         0       2    male  47.0      0      0   15.0000        S  Second   
592         0       3    male  47.0      0      0    7.2500        S   Third   
597         0       3    male  49.0      0      0    0.0000        S   Third   
599         1       1    male  49.0      1      0   56.9292        C   First   
645         1       1    male  48.0      1      0   76.7292        C   First   
662         0       1    male  47.0      0      0   25.5875        S   First   
698         0       1    male  49.0      1      1  110.8833        C   First   
712         1       1    male  48.0      1      0   52.0000        S   First   
736         0       3  female  48.0      1      3   34.3750        S   Third   
754         1       2  female  48.0      1      2   65.0000        S  Second   
771         0       3    male  48.0      0      0    7.8542        S   Third   
796         1       1  female  49.0      0      0   25.9292        S   First   
862         1       1  female  48.0      0      0   25.9292        S   First   
871         1       1  female  47.0      1      1   52.5542        S   First   
873         0       3    male  47.0      0      0    9.0000        S   Third   

     adult_male deck  embark_town alive  alone  
52        False    D    Cherbourg   yes  False  
110        True    C  Southampton    no   True  
132       False  NaN  Southampton    no  False  
453        True    C    Cherbourg   yes  False  
460        True    E  Southampton   yes   True  
462        True    E  Southampton    no   True  
463        True  NaN  Southampton    no   True  
515        True    D  Southampton    no   True  
556       False    A    Cherbourg   yes  False  
586        True  NaN  Southampton    no   True  
592        True  NaN  Southampton    no   True  
597        True  NaN  Southampton    no   True  
599        True    A    Cherbourg   yes  False  
645        True    D    Cherbourg   yes  False  
662        True    E  Southampton    no   True  
698        True    C    Cherbourg    no  False  
712        True    C  Southampton   yes  False  
736       False  NaN  Southampton    no  False  
754       False  NaN  Southampton   yes  False  
771        True  NaN  Southampton    no   True  
796       False    D  Southampton   yes   True  
862       False    D  Southampton   yes   True  
871       False    D  Southampton   yes  False  
873        True  NaN  Southampton    no   True

- - -
8 - Obten los nombres de los pueblos de embarcación de los pasajeros de la cubierta "A"

`Out:
23     Southampton
96       Cherbourg
174      Cherbourg
185    Southampton
209      Cherbourg
284    Southampton
445    Southampton
475    Southampton
556      Cherbourg
583      Cherbourg
599      Cherbourg
630    Southampton
647      Cherbourg
806    Southampton
867    Southampton
Name: embark_town, dtype: object`

In [99]:
df[df["deck"].isin(["A"])] # filtra el dataframe por los que tienen exactamente la palabra "A" en el campo "deck"

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
23          1       1    male  28.0      0      0  35.5000        S  First   
96          0       1    male  71.0      0      0  34.6542        C  First   
174         0       1    male  56.0      0      0  30.6958        C  First   
185         0       1    male   NaN      0      0  50.0000        S  First   
209         1       1    male  40.0      0      0  31.0000        C  First   
284         0       1    male   NaN      0      0  26.0000        S  First   
445         1       1    male   4.0      0      2  81.8583        S  First   
475         0       1    male   NaN      0      0  52.0000        S  First   
556         1       1  female  48.0      1      0  39.6000        C  First   
583         0       1    male  36.0      0      0  40.1250        C  First   
599         1       1    male  49.0      1      0  56.9292        C  First   
630         1       1    male  80.0      0      0  30.0000        S  First   
647         1       1    male  56.0      0      0  35.5000        C  First   
806         0       1    male  39.0      0      0   0.0000        S  First   
867         0       1    male  31.0      0      0  50.4958        S  First   

     adult_male deck  embark_town alive  alone  
23         True    A  Southampton   yes   True  
96         True    A    Cherbourg    no   True  
174        True    A    Cherbourg    no   True  
185        True    A  Southampton    no   True  
209        True    A    Cherbourg   yes   True  
284        True    A  Southampton    no   True  
445       False    A  Southampton   yes  False  
475        True    A  Southampton    no   True  
556       False    A    Cherbourg   yes  False  
583        True    A    Cherbourg    no   True  
599        True    A    Cherbourg   yes  False  
630        True    A  Southampton   yes   True  
647        True    A    Cherbourg   yes   True  
806        True    A  Southampton    no   True  
867        True    A  Southampton    no   True

In [100]:
df[df["deck"].isin(["A"])]["embark_town"] # Para tener la columna de los pueblos de embarcación

23     Southampton
96       Cherbourg
174      Cherbourg
185    Southampton
209      Cherbourg
284    Southampton
445    Southampton
475    Southampton
556      Cherbourg
583      Cherbourg
599      Cherbourg
630    Southampton
647      Cherbourg
806    Southampton
867    Southampton
Name: embark_town, dtype: object

- - -

9 - Muestra la media de los valores del dataframe agrupando por sexo

In [101]:
df.groupby("sex").mean()

<ipython-input-101-4636136becc0>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby("sex").mean()


survived    pclass        age     sibsp     parch       fare  \
sex                                                                    
female  0.742038  2.159236  27.915709  0.694268  0.649682  44.479818   
male    0.188908  2.389948  30.726645  0.429809  0.235702  25.523893   

        adult_male     alone  
sex                           
female    0.000000  0.401274  
male      0.930676  0.712305

- - -
10 - Define una función que capitalice strings y con pandas aplicala a la columna sexo

`Entrada : hola mundo`
`Salida: Hola mundo`
`Pista: capitalize()`

In [102]:
def capital(s):
  return(s.capitalize())
df["sex"].apply(capital)

0        Male
1      Female
2      Female
3      Female
4        Male
        ...  
886      Male
887    Female
888    Female
889      Male
890      Male
Name: sex, Length: 891, dtype: object

11 - Reemplaza todos los NaN de la columna deck por la palabra 'Unknown'

In [103]:
df["deck"].fillna("Unknown")

0      Unknown
1            C
2      Unknown
3            C
4      Unknown
        ...   
886    Unknown
887          B
888    Unknown
889          C
890    Unknown
Name: deck, Length: 891, dtype: object

12 - ¿Cuál es el precio promedio del billete que pagaron los pasajeros por cada clase (primera, segunda, tercera)?

`Out: First     84.154687
Second    20.662183
Third     13.675550 `

In [104]:
df.groupby("class").mean().drop(["survived","pclass","age", "sibsp","parch","adult_male","alone"], axis=1)

<ipython-input-104-3a71af035abf>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby("class").mean().drop(["survived","pclass","age", "sibsp","parch","adult_male","alone"], axis=1)


fare
class            
First   84.154687
Second  20.662183
Third   13.675550

- - -

13 - ¿Cuántos supervivientes están todavía vivos?


 `Out: 342`

In [105]:
df["survived"].sum() # Dado que 1=sobrevivió y 0=no sobrevivió, si sumamos los valores de esa columna sabremos cuantos sobrevivieron

342

In [106]:
df["survived"].fillna(0).sum() # Hacemos lo mismo suponiendo que los que no sabemos si sobrevivieron, murieron

342

In [107]:
df["survived"].fillna(1).sum() # Hacemos lo mismo suponiendo que los que no sabemos si sobrevivieron, sobrevivieron

342

- - -

14 - ¿Cuál fue el precio del billete de segunda clase más caro? ¿Y el más barato?


`Out : Caro: 73.5 Barato: 0.0 `

In [108]:
df[df["class"].isin(["Second"])] # filtra el dataframe por los que tienen exactamente la palabra "Second" en el campo "class"

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
9           1       2  female  14.0      1      0  30.0708        C  Second   
15          1       2  female  55.0      0      0  16.0000        S  Second   
17          1       2    male   NaN      0      0  13.0000        S  Second   
20          0       2    male  35.0      0      0  26.0000        S  Second   
21          1       2    male  34.0      0      0  13.0000        S  Second   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
866         1       2  female  27.0      1      0  13.8583        C  Second   
874         1       2  female  28.0      1      0  24.0000        C  Second   
880         1       2  female  25.0      0      1  26.0000        S  Second   
883         0       2    male  28.0      0      0  10.5000        S  Second   
886         0       2    male  27.0      0      0  13.0000        S  Second   

     adult_male deck  embark_town alive  alone  
9         False  NaN    Cherbourg   yes  False  
15        False  NaN  Southampton   yes   True  
17         True  NaN  Southampton   yes   True  
20         True  NaN  Southampton    no   True  
21         True    D  Southampton   yes   True  
..          ...  ...          ...   ...    ...  
866       False  NaN    Cherbourg   yes  False  
874       False  NaN    Cherbourg   yes  False  
880       False  NaN  Southampton   yes  False  
883        True  NaN  Southampton    no   True  
886        True  NaN  Southampton    no   True  

[184 rows x 14 columns]

In [109]:
#seleccionamos solo valores de precio billete
select = df[df["class"].isin(["Second"])].drop(["class","sex","survived","pclass","age", "sibsp","parch","adult_male","alone", "embarked",	"deck",	"embark_town","alive"], axis=1)
select

fare
9    30.0708
15   16.0000
17   13.0000
20   26.0000
21   13.0000
..       ...
866  13.8583
874  24.0000
880  26.0000
883  10.5000
886  13.0000

[184 rows x 1 columns]

In [110]:
# Encontramos el máximo de allí
select.max()

fare    73.5
dtype: float64

In [111]:
# Encontramos el mínimo de allí
select.min()

fare    0.0
dtype: float64

- - -

15- Elimina el tercer registro del dataframe

- - -

In [112]:
df.drop(2) # Quitamos el tercer registro (fila 2)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
5           0       3    male   NaN      0      0   8.4583        Q   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

     adult_male deck  embark_town alive  alone  
0          True  NaN  Southampton    no  False  
1         False    C    Cherbourg   yes  False  
3         False    C  Southampton   yes  False  
4          True  NaN  Southampton    no   True  
5          True  NaN   Queenstown    no   True  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[890 rows x 14 columns]

In [113]:
df.drop(3) # Quitamos la fila 3

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
4           0       3    male  35.0      0      0   8.0500        S   Third   
5           0       3    male   NaN      0      0   8.4583        Q   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

     adult_male deck  embark_town alive  alone  
0          True  NaN  Southampton    no  False  
1         False    C    Cherbourg   yes  False  
2         False  NaN  Southampton   yes   True  
4          True  NaN  Southampton    no   True  
5          True  NaN   Queenstown    no   True  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[890 rows x 14 columns]

16- Añade un nuevo registro al dataframe de un hombre que embarcó desde el puerto de Barcelona. Puedes inventarte el resto de campos.

In [114]:
nueva_fila = ["1", "3", "male", "28", "3", "2", "7.7500", "S", "Third", True, "C", "Barcelona", "yes", False]
#if not ( df.iloc[[1]].all() == nueva_fila ):
#  df.loc[len(df)] = nueva_fila
df.loc[len(df)] = nueva_fila
df.tail()

survived pclass     sex   age sibsp parch    fare embarked  class  \
887        1      1  female  19.0     0     0    30.0        S  First   
888        0      3  female   NaN     1     2   23.45        S  Third   
889        1      1    male  26.0     0     0    30.0        C  First   
890        0      3    male  32.0     0     0    7.75        Q  Third   
891        1      3    male    28     3     2  7.7500        S  Third   

     adult_male deck  embark_town alive  alone  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  
891        True    C    Barcelona   yes  False

- - -

17- Ordena el dataframe según la edad de los registros.

---



In [115]:
df.sort_values(by='age')

TypeError: '<' not supported between instances of 'str' and 'float'

18 - Guarda el dataframe como un fichero de Excel.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
df.to_excel("/content/drive/MyDrive/Titanic_Albert_Ripoll.xlsx", index=False, sheet_name="Hoja1")

- - -

# Pandas - Ejercicios dataset Youtube

1- Descárgate el dataset de YouTube del eStudy y cárgalo en un DataFrame.

In [117]:
import pandas as pd
# Ponemos el documento "youtube_dataset.csv" en google Collab en la carpeta del menú de la izquierda.
df = pd.read_csv('youtube_dataset.csv')

2- Muestra los 3 primeros registros.

In [118]:
df.head(3)

Rank                                 Name Brand channel  \
0     1                             T-Series           Yes   
1     2                            Cocomelon           Yes   
2     3  Sony Entertainment Television India           Yes   

   Subscribers (millions) Primary language       Category         Country  
0                   232.0      Hindi[7][8]          Music           India  
1                   150.0          English      Education   United States  
2                   149.0         Hindi[9]  Entertainment           India

3- ¿Cuál es el promedio de suscriptores que tiene cada canal?

In [119]:
df["Subscribers (millions)"].isna().sum()

0

In [120]:
df["Subscribers (millions)"].mean()

70.442

4- ¿Qué canal tiene el nombre más largo?

In [121]:
# apply(len) calcula la longitud de cada elemento de la columna "Name"
df["Name"].apply(len)

0      8
1      9
2     35
3      7
4      9
5     15
6     11
7      3
8     13
9     17
10     9
11     9
12    15
13     8
14     9
15    13
16    11
17    15
18     6
19     8
20    22
21     9
22     9
23    12
24    10
25    21
26    10
27    15
28    10
29    16
30     7
31     6
32    17
33    10
34    13
35    11
36    12
37    14
38    19
39     9
40    13
41    11
42     7
43    10
44     9
45    11
46    16
47    11
48    11
49    17
Name: Name, dtype: int64

In [122]:
# idxmax() devuelve el índice del elemento con la longitud máxima
df["Name"].apply(len).idxmax()

2

In [123]:
# df.loc obtiene la fila de ese índice
df.loc[indice_max_longitud]

Rank                                                        3
Name                      Sony Entertainment Television India
Brand channel                                             Yes
Subscribers (millions)                                  149.0
Primary language                                     Hindi[9]
Category                                        Entertainment
Country                                                 India
Name: 2, dtype: object

5- ¿Qué porcentaje de canales son de marcas?

In [124]:
df["Brand channel"].value_counts()

Yes    31
No     19
Name: Brand channel, dtype: int64

In [125]:
df["Brand channel"].value_counts("Yes")

Yes    0.62
No     0.38
Name: Brand channel, dtype: float64

In [126]:
df["Brand channel"].value_counts("Yes")[0]*100

62.0

6- ¿Cuál es la categoría menos popular? ¿Y la más popular?

In [127]:
df["Category"].value_counts()

Music                     24
Entertainment             13
Education                  5
Film                       3
Professional wrestling     1
How-to                     1
Sports                     1
News                       1
Gaming                     1
Name: Category, dtype: int64

In [128]:
df["Category"].value_counts().idxmax()

'Music'

7- ¿Cuántos canales Educativos tienen el Inglés como lengua principal?

In [129]:
df["Primary language"].value_counts()

English                20
Hindi                  12
Spanish                 5
Portuguese              4
Korean                  3
Hindi[7][8]             1
Hindi[9]                1
English[10][11][12]     1
Hindi[13][14]           1
Hindi[16]               1
Bhojpuri                1
Name: Primary language, dtype: int64

In [130]:
df["Primary language"].value_counts()["English"] + df["Primary language"].value_counts()["English[10][11][12]"]

21

8- Crea una nueva columna en el dataset con el continente de cada canal.

In [131]:
df["Country"].value_counts()

 India                           17
 United States                   12
 South Korea                      4
 Brazil                           4
 Cyprus[a]                        2
 Sweden                           1
 Ukraine                          1
 Russia                           1
 Canada                           1
 Argentina                        1
 United Kingdom                   1
 Romania                          1
 Chile                            1
 Mexico                           1
 El Salvador                      1
 United States ( Puerto Rico)     1
Name: Country, dtype: int64

In [132]:
# Define a dictionary mapping countries to continents
country_to_continent = {
    'India': 'Asia',
    'United States': 'North America',
    'South Korea': 'Asia',
    'Brazil': 'South America',
    'Cyprus': 'Europe',
    'Sweden': 'Europe',
    'Ukraine': 'Europe',
    'Russia': 'Europe',
    'Canada': 'North America',
    'Argentina': 'South America',
    'United Kingdom': 'Europe',
    'Romania': 'Europe',
    'Chile': 'South America',
    'Mexico': 'North America',
    'El Salvador': 'North America',
    'United States (Puerto Rico)': 'North America'
}

# Clean spaces, dots, * and slashes that might be the 'Country' column in the DataFrame to match the dictionary
df['Country'] = df['Country'].str.replace(r'\[.*\]', '', regex=True).str.strip()

# Create a new 'Continent' column by mapping the values from the 'Country' column
df['Continent'] = df['Country'].map(country_to_continent)
df.head(3)

Rank                                 Name Brand channel  \
0     1                             T-Series           Yes   
1     2                            Cocomelon           Yes   
2     3  Sony Entertainment Television India           Yes   

   Subscribers (millions) Primary language       Category        Country  \
0                   232.0      Hindi[7][8]          Music          India   
1                   150.0          English      Education  United States   
2                   149.0         Hindi[9]  Entertainment          India   

       Continent  
0           Asia  
1  North America  
2           Asia

9- Enumera el promedio de suscriptores por continente.

In [133]:
df.groupby('Continent')['Subscribers (millions)'].mean()

Continent
Asia             75.400000
Europe           76.557143
North America    70.513333
South America    50.050000
Name: Subscribers (millions), dtype: float64

10- Crea un DataFrame independiente con la información de los canales de YouTube de Lady Gaga y La Salle Campus Barcelona que tenga como datos el nombre del canal, los suscriptores y la categoría.

In [136]:
data = {'Name': ['Lady Gaga', 'La Salle Campus Barcelona'],
        'Subscribers (millions)': [50.0, 0.1],
        'Categoría': ['Music', 'Education']}
df_2 = pd.DataFrame(data)

11- Une el DataFrame creado en el punto 10 con el DataFrame original en un nuevo DataFrame.

In [137]:
pd.merge(df, df_2, how="outer")

Rank                                 Name Brand channel  \
0    1.0                             T-Series           Yes   
1    2.0                            Cocomelon           Yes   
2    3.0  Sony Entertainment Television India           Yes   
3    4.0                              MrBeast            No   
4    5.0                            PewDiePie            No   
5    6.0                      Kids Diana Show           Yes   
6    7.0                          Like Nastya            No   
7    8.0                                  WWE           Yes   
8    9.0                        Vlad and Niki            No   
9   10.0                    Zee Music Company           Yes   
10  11.0                            Blackpink           Yes   
11  12.0                            Goldmines           Yes   
12  13.0                      5-Minute Crafts           Yes   
13  14.0                             Sony SAB           Yes   
14  15.0                            BangtanTV            No   
15  16.0                        Justin Bieber            No   
16  17.0                          Hybe Labels           Yes   
17  18.0                      Canal KondZilla           Yes   
18  19.0                               Zee TV           Yes   
19  20.0                             Pinkfong           Yes   
20  21.0               Shemaroo Entertainment           Yes   
21  22.0                            ChuChu TV           Yes   
22  23.0                            Colors TV           Yes   
23  24.0                         Dude Perfect            No   
24  25.0                           Movieclips           Yes   
25  26.0                T-Series Bhakti Sagar           Yes   
26  27.0                           Marshmello            No   
27  28.0                      Tips Industries           Yes   
28  29.0                           Wave Music           Yes   
29  30.0                     Sony Music India           Yes   
30  31.0                              Aaj Tak           Yes   
31  32.0                               Eminem            No   
32  33.0                    El Reino Infantil           Yes   
33  34.0                           Ed Sheeran            No   
34  35.0                        Ariana Grande            No   
35  36.0                          LooLoo Kids           Yes   
36  37.0                         Taylor Swift            No   
37  38.0                       Yash Raj Films           Yes   
38  39.0                  BillionSurpriseToys           Yes   
39  40.0                            Infobells           Yes   
40  41.0                        Billie Eilish            No   
41  42.0                          JuegaGerman            No   
42  43.0                              Badabun           Yes   
43  44.0                           Fernanfloo            No   
44  45.0                            Bad Bunny            No   
45  46.0                          Felipe Neto            No   
46  47.0                     Sony Music India           Yes   
47  48.0                          Bright Side           Yes   
48  49.0                          Você Sabia?            No   
49  50.0                    Whindersson Nunes            No   
50   NaN                            Lady Gaga           NaN   
51   NaN            La Salle Campus Barcelona           NaN   

    Subscribers (millions)     Primary language                Category  \
0                    232.0          Hindi[7][8]                   Music   
1                    150.0              English               Education   
2                    149.0             Hindi[9]           Entertainment   
3                    124.0              English           Entertainment   
4                    111.0              English           Entertainment   
5                    106.0  English[10][11][12]                    Film   
6                    103.0              English           Entertainment   
7                     92.4              English  Professional wr

12 - Modifica los nombres de los canales para que no haya espacio en ellos. Para ello, reemplaza todos los espacios por el carácter "_".

In [140]:
df['Name'] = df['Name'].str.replace(' ', '_')
df.head(3)

Rank                                 Name Brand channel  \
0     1                             T-Series           Yes   
1     2                            Cocomelon           Yes   
2     3  Sony_Entertainment_Television_India           Yes   

   Subscribers (millions) Primary language       Category        Country  \
0                   232.0      Hindi[7][8]          Music          India   
1                   150.0          English      Education  United States   
2                   149.0         Hindi[9]  Entertainment          India   

       Continent  
0           Asia  
1  North America  
2           Asia